### PR TITLE
Relation backlinks

### PR DESCRIPTION
**Pull request by vaind**
Part of #90 

* [x] backlink to ToOne
* [x] backlink to ToMany
* [x] backlinks in queries

### CODE REVIEW

File: lib/src/modelinfo/modelproperty.dart [hunk 1]
```diff
@@ -98,6 +98,8 @@ class ModelProperty {
         hasFlag(OBXPropertyFlags.INDEX_HASH64);
   }
 
+  bool get isRelation => type == OBXPropertyType.Relation;
```
Comment (greenrobot-team) [inline lib/src/modelinfo/modelproperty.dart:101]: The `prop.type == OBXPropertyType.Relation` check is still used in some other places. Should these be replaced as well?

File: lib/src/relations/to_many.dart [hunk 2]
```diff
@@ -141,8 +143,7 @@ class ToMany<EntityT> extends Object with ListMixin<EntityT> {
       _items.toList(growable: growable);
 
   /// True if there are any changes not yet saved in DB.
-  bool get _hasPendingDbChanges =>
-      _counts.values.where((c) => c != 0).isNotEmpty;
+  bool get _hasPendingDbChanges => _counts.values.any((c) => c != 0);
 
   // bool get _hasPendingDbChanges => _added.isNotEmpty || _removed.isNotEmpty;
```
Comment (greenrobot-team) [inline lib/src/relations/to_many.dart:148]: Just noticed this leftover comment, can this be removed?

```suggestion
```

File: test/relations_test.dart [hunk 3]
```diff
@@ -300,6 +300,140 @@ void main() {
...
+      expect(a[1].testEntities.map(strings), unorderedEquals(['foo', 'bar2']));
+      boxA.put(a[1]);
+      a[1] = boxA.get(a[1].id);
+      expect(a[1].testEntities.map(strings), unorderedEquals(['foo', 'bar2']));
+
+      // Insert a new target, already with some "source" entities pointing to it.
+      var newA = RelatedEntityA(tInt: 4);
+      expect(newA.testEntities.length, isZero);
+      newA.testEntities.add(env.box.get(1)); // foo
+      newA.testEntities.add(TestEntity(tString: 'newly created from B'));
```
Comment (greenrobot-team) [inline test/relations_test.dart:412]: ?
```suggestion
      newA.testEntities.add(TestEntity(tString: 'newly created from A'));
```

File: test/relations_test.dart [hunk 4]
```diff
@@ -300,6 +300,140 @@ void main() {
...
+      boxA.put(newA);
+      expect(newA.testEntities[0].id, 1);
+      expect(newA.testEntities[1].id, 4);
+
+      expect(env.box.get(4).tString, equals('newly created from B'));
```
Comment (greenrobot-team) [inline test/relations_test.dart:417]: ?
```suggestion
      expect(env.box.get(4).tString, equals('newly created from A'));
```

File: test/relations_test.dart [hunk 5]
```diff
@@ -300,6 +300,140 @@ void main() {
...
+      newA = boxA.get(newA.id);
+      expect(newA.testEntities.map(strings),
+          unorderedEquals(['foo', 'newly created from B']));
```
Comment (greenrobot-team) [inline test/relations_test.dart:420]: ?
```suggestion
          unorderedEquals(['foo', 'newly created from A']));
```

**Review by greenrobot-team — COMMENTED** [review]
LGTM, potentially just some minor clean up stuff.

File: lib/src/modelinfo/modelproperty.dart (hunk 1 above)
Comment (vaind) [reply lib/src/modelinfo/modelproperty.dart:101]: replaced

File: lib/src/relations/to_many.dart (hunk 2 above)
Comment (vaind) [reply lib/src/relations/to_many.dart:148]: removed

File: test/relations_test.dart (hunk 4 above)
Comment (vaind) [reply test/relations_test.dart:417]: fixed

**Pull request merged**